rclcpp/include/rclcpp/rate.hpp

```diff
@@ -112,8 +120,109 @@ class GenericRate : public RateBase
   std::chrono::time_point<Clock, ClockDurationNano> last_interval_;
 };
 
-using Rate = GenericRate<std::chrono::system_clock>;
-using WallRate = GenericRate<std::chrono::steady_clock>;
+class Rate : public RateBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(Rate)
+
+  explicit Rate(
+    const double rate,
+    Clock::SharedPtr clock = std::make_shared<Clock>(RCL_SYSTEM_TIME))
+  : Rate(
+      Duration::from_seconds(1.0 / rate), clock)
+  {}
+  explicit Rate(
+    const Duration & period,
+    Clock::SharedPtr clock = std::make_shared<Clock>(RCL_SYSTEM_TIME))
+  : clock_(clock), period_(period), last_interval_(clock_->now())
+  {}
+
+  virtual bool
+  sleep()
+  {
+    // Time coming into sleep
+    auto now = clock_->now();
+    // Time of next interval
+    auto next_interval = last_interval_ + period_;
+    // Detect backwards time flow
+    if (now < last_interval_) {
+      // Best thing to do is to set the next_interval to now + period
+      next_interval = now + period_;
+    }
+    // Update the interval
+    last_interval_ += period_;
+    // If the time_to_sleep is negative or zero, don't sleep
+    if (next_interval <= now) {
+      // If an entire cycle was missed then reset next interval.
+      // This might happen if the loop took more than a cycle.
+      // Or if time jumps forward.
+      if (now > next_interval + period_) {
+        last_interval_ = now + period_;
+      }
+      // Either way do not sleep and return false
+      return false;
+    }
+    // Calculate the time to sleep
+    auto time_to_sleep = next_interval - now;
+    // Sleep (will get interrupted by ctrl-c, may not sleep full time)
+    return clock_->sleep_for(time_to_sleep);
+  }
```

Comment:
Given that this class is not a template, I'd like to see all of this implementation moved into a `rate.cpp` file.  That will help with compile times of downstream projects.

The same goes for all of the other methods below.

Reply:
Moved all `Rate` implementations into source-file in https://github.com/ros2/rclcpp/pull/2123/commits/3d7b2232dca56ddd1021c174a4617714bb2f9538